--- linux-cc	2012-05-09 06:01:41.000000000 -0700
+++ bin/linux-cc	2012-08-09 12:58:37.283453567 -0700
@@ -1701,6 +1701,35 @@ else
      LIBPNETCDF=
 fi
 
+###
+# Process Darshan libraries
+###
+
+if [ ${CRAY_DARSHAN_DIR:+1} ] ; then
+
+   unset LIBDARSHAN
+   unset DARSHAN_POST_LINK_OPTS
+
+   # disable darshan if dynamic linking
+   if [ "$XTPE_LINK_TYPE" != "dynamic" ] ; then
+
+      case $PE_ENV in     
+      PGI|GNU|PATHSCALE|INTEL )
+         LIBDARSHAN=`${CRAY_DARSHAN_DIR}/bin/darshan-config --post-ld-flags`
+         DARSHAN_POST_LINK_OPTS=`${CRAY_DARSHAN_DIR}/bin/darshan-config --pre-ld-flags`
+      ;;
+      *)
+         # TODO: test with Cray environment
+      ;;
+      esac
+
+   fi
+
+else
+   LIBDARSHAN=
+   DARSHAN_POST_LINK_OPTS=
+fi
+
 ##
 # PMI
 ##
@@ -1979,6 +2008,11 @@ fi
         eval $($BUILD_OPTS)
    fi
 
+if [ ${DARSHAN_POST_LINK_OPTS:+1} ] ; then
+   CNL_LIBS_LIST="$CNL_LIBS_LIST $LIBDARSHAN"
+   POST_COMPILE_OPTS="${DARSHAN_POST_LINK_OPTS} ${POST_LINK_OPTS}" 
+fi
+
 comp_only=0
 
 for arg
