*** include.org/darshan-logutils.h	2018-06-30 18:43:37.811199360 -0400
--- include/darshan-logutils.h	2018-06-30 18:53:02.536449338 -0400
***************
*** 7,19 ****
  #ifndef __DARSHAN_LOG_UTILS_H
  #define __DARSHAN_LOG_UTILS_H
  
- #include <limits.h>
- #include <zlib.h>
- #ifdef HAVE_LIBBZ2
- #include <bzlib.h>
- #endif
- 
- #include "uthash-1.9.2/src/uthash.h"
  
  #include "darshan-log-format.h"
  
--- 7,12 ----
