*** generated.h.org	2018-06-30 18:16:31.104359748 -0400
--- generated.h	2018-06-30 18:18:46.285180237 -0400
***************
*** 305,310 ****
--- 305,311 ----
  };
  typedef struct darshan_fd_s* darshan_fd;
  
+ /*
  struct darshan_name_record_ref
  {
      struct darshan_name_record *name_record;
***************
*** 317,327 ****
   struct darshan_lustre_record *rec;
   UT_hash_handle hlink;
  };
  
  struct darshan_mnt_info
  {
!     char mnt_type[(4096 - sizeof(struct darshan_job) - 1)];
!     char mnt_path[(4096 - sizeof(struct darshan_job) - 1)];
  };
  
  
--- 318,329 ----
   struct darshan_lustre_record *rec;
   UT_hash_handle hlink;
  };
+ */
  
  struct darshan_mnt_info
  {
! 		char mnt_type[3031];
! 		char mnt_path[3031];
  };
  
  
